File: wear/src/main/java/technology/mainthread/apps/gatekeeper/messenger/WearToMobileRequester.java
package technology.mainthread.apps.gatekeeper.messenger;

import javax.inject.Inject;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.FlowableOnSubscribe;
import technology.mainthread.apps.gatekeeper.common.SharedValues;

public class WearToMobileRequester {

    private WearableMessenger messenger;

    @Inject
    public WearToMobileRequester(WearableMessenger messenger) {
        this.messenger = messenger;
    }

    public Flowable<Boolean> prime() {
        FlowableOnSubscribe<Boolean> source = e -> {
            e.onNext(messenger.sendMessage(SharedValues.PATH_PRIME, new byte[0]));
            e.onComplete();
        };
        return Flowable.create(source, BackpressureStrategy.BUFFER);
    }

    public Flowable<Boolean> unlock() {
        FlowableOnSubscribe<Boolean> source = e -> {
            e.onNext(messenger.sendMessage(SharedValues.PATH_UNLOCK, new byte[0]));
            e.onComplete();
        };
        return Flowable.create(source, BackpressureStrategy.BUFFER);
    }
}

File: wear/src/main/java/technology/mainthread/apps/gatekeeper/messenger/WearToMobileRequester.kt
package technology.mainthread.apps.gatekeeper.messenger

import io.reactivex.BackpressureStrategy
import io.reactivex.Emitter
import io.reactivex.Flowable
import technology.mainthread.apps.gatekeeper.common.PATH_PRIME
import technology.mainthread.apps.gatekeeper.common.PATH_UNLOCK
import javax.inject.Inject

class WearToMobileRequester @Inject
constructor(private val messenger: WearableMessenger) {

    fun prime(): Flowable<Boolean> {
        val source = { e: Emitter<Boolean> ->
            e.onNext(messenger.sendMessage(PATH_PRIME, ByteArray(0)))
            e.onComplete()
        }
        return Flowable.create<Boolean>(source, BackpressureStrategy.BUFFER)
    }

    fun unlock(): Flowable<Boolean> {
        val source = { e: Emitter<Boolean> ->
            e.onNext(messenger.sendMessage(PATH_UNLOCK, ByteArray(0)))
            e.onComplete()
        }
        return Flowable.create<Boolean>(source, BackpressureStrategy.BUFFER)
    }
}
